src/client/utils/timeoutMiddleware.ts

```diff
@@ -0,0 +1,93 @@
+import type { JSONRPCRequest, JSONRPCResponse } from '../../rpc/types';
+import type { ClientRPCRequestParams, ClientRPCResponseResult } from '../types';
+import type { ContextTimed } from '../../contexts/types';
+import { TransformStream } from 'stream/web';
+
+/**
+ * This adds its timeout to the reverse metadata and updates it's timeout based
+ * on the forward metadata.
+ */
+function timeoutMiddlewareServer(ctx: ContextTimed) {
```

Comment:
This is going to need to be abstracted out to a common middleware location. That could be `src/rpc/middleware`, that could be `src/middleware`.

I can see here you are using `ClientRPCRequestParams` here, but this could be a generic type parameter.

```
timeoutMiddlewareServer<Req, Res> {
  ...
}
```

Furthermore the generic timeout middleware should work even for raw streams. This is why I mentioned you should be able to reset the timer upon any data chunk.

Reply:
So because the transport layer already has timeout for idleness that tracks bytes being transferred, there's no need to have a timeout for raw streams.

Therefore in this scenario, this is a "RPC" level timeout which has to involve JSON because you have the `timeout` property in the params.

Reply:
So that means we have 2 layers of timeout.

The bottom layer is concerned with idleness at the networking packet level.

The top layer is concerned with application execution of the handlers.

Reply:
So the only thing needed here is to figure out how to abstract these generic middlewares between `client` and `agent` domains.

Reply:
I'll add some details to the agent migration issue. this will be addressed there.

Reply:
I mentioned some details in the agent migration issue #512